sdk/python/sawtooth_sdk/client/encoding.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2017 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+from hashlib import sha512
+from random import randint
+from google.protobuf.message import DecodeError
+
+from sawtooth_sdk.client import signing
+from sawtooth_sdk.protobuf.batch_pb2 import Batch
+from sawtooth_sdk.protobuf.batch_pb2 import BatchHeader
+from sawtooth_sdk.protobuf.batch_pb2 import BatchList
+from sawtooth_sdk.protobuf.transaction_pb2 import Transaction
+from sawtooth_sdk.protobuf.transaction_pb2 import TransactionHeader
+from sawtooth_sdk.protobuf.transaction_pb2 import TransactionList
+
+
+def _listify(item_or_items):
+    """Wraps an item in a list if it is not already a list
+    """
+    if isinstance(item_or_items, list):
+        return item_or_items
+    return [item_or_items]
+
+
+class TransactionEncoder(object):
+    """Stores a private key and default TransactionHeader values, which will
+    then be used to create and serialize Transactions. Any TransactionHeader
+    value set will be used with EVERY Transaction this encoder creates.
+
+    Args:
+        private_key (str or bytes): The private key to sign Transactions with
+        payload_encoder (function, optional): Run on each payload before
+            creating Transaction, must return bytes
+        batcher_pubkey (string or bytes, optional): Expected batcher public key
+        dependencies (list of str, optional): Transaction ids that must be
+            committed before every Transaction from this encoder (unusual)
+        family_name (str, optional): Name of the designated Transaction Family
+        family_version (str, optional): Version of Transaction Family
+        inputs (list of str, optional): State addresses that every Transaction
+            created by this encoder will read from (unusual)
+        outputs (list of str, optional): State addresses that every Transaction
+            created by this encoder will write to (unusual)
+        payload_encoding (str, optional): The byte-encoding the Transaction
+            Processor should expect
+
+    Note:
+        Every optional argument can be set or changed later as an attribute.
+    """
+    def __init__(self,
+                 private_key,
+                 payload_encoder=lambda x: x,
+                 batcher_pubkey=None,
+                 dependencies=None,
+                 family_name=None,
```

Comment:
The TransactionEncoder tries to be unopinionated. Most people will probably set those three defaults, and only those three, and not override them. But it is entirely possible that someone might create an encoder and use it with multiple families or what have you.